## Part 1c: TensorFlow - Early Stopping

**Description:**

This Colab demonstrates the use of the Early Stopping callback in TensorFlow Keras. Early stopping is a regularization technique that monitors the model's performance on a validation set and stops training when the performance stops improving for a specified number of epochs. This prevents overfitting by halting the training process before the model starts to memorize the training data.

In [1]:
import tensorflow as tf
from tensorflow.keras import layers, models
from tensorflow.keras.callbacks import EarlyStopping
from sklearn.model_selection import train_test_split
from sklearn.datasets import load_digits
from sklearn.preprocessing import StandardScaler

# Load the digits dataset
digits = load_digits()
X, y = digits.data, digits.target

# Scale the data
scaler = StandardScaler()
X_scaled = scaler.fit_transform(X)

# Split the data into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X_scaled, y, test_size=0.2, random_state=42)

# Define the EarlyStopping callback
early_stopping = EarlyStopping(monitor='val_loss', patience=10, restore_best_weights=True)

# Model without Early Stopping
model_no_es = models.Sequential([
    layers.Dense(64, activation='relu', input_shape=(X_train.shape[1],)),
    layers.Dense(10, activation='softmax')
])
model_no_es.compile(optimizer='adam', loss='sparse_categorical_crossentropy', metrics=['accuracy'])
history_no_es = model_no_es.fit(X_train, y_train, epochs=100, validation_data=(X_test, y_test), verbose=0)

# Model with Early Stopping
model_early_stopping = models.Sequential([
    layers.Dense(64, activation='relu', input_shape=(X_train.shape[1],)),
    layers.Dense(10, activation='softmax')
])
model_early_stopping.compile(optimizer='adam', loss='sparse_categorical_crossentropy', metrics=['accuracy'])
history_early_stopping = model_early_stopping.fit(
    X_train, y_train,
    epochs=100,  # Train for more epochs to allow early stopping to activate
    validation_data=(X_test, y_test),
    callbacks=[early_stopping],
    verbose=0
)

# Evaluate the models
print("\nEvaluation:")
print("No Early Stopping - Test Accuracy:", model_no_es.evaluate(X_test, y_test, verbose=0)[1])
print("Early Stopping - Test Accuracy:", model_early_stopping.evaluate(X_test, y_test, verbose=0)[1])

/usr/local/lib/python3.11/dist-packages/keras/src/layers/core/dense.py:87: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)



Evaluation:
No Early Stopping - Test Accuracy: 0.9777777791023254
Early Stopping - Test Accuracy: 0.9861111044883728


## Results for Part 1c: TensorFlow - Early Stopping

In this experiment, we investigated the impact of using the Early Stopping callback during the training of a simple neural network on the `digits` dataset. We compared a model trained for a fixed number of epochs (100 in this case) with a model that used Early Stopping based on the validation loss. The Early Stopping callback was configured to monitor the validation loss (`val_loss`) with a patience of 10 epochs and to restore the best weights found during training.

The test accuracies achieved by each model are as follows:

* **No Early Stopping - Test Accuracy:** 0.9778
* **Early Stopping - Test Accuracy:** 0.9861

**Analysis:**

The results clearly demonstrate the benefit of using Early Stopping in this scenario:

* The model trained with **Early Stopping** achieved a higher test accuracy of 0.9861 compared to the model trained for a fixed number of epochs, which reached 0.9778.

This improvement suggests that Early Stopping effectively prevented overfitting by stopping the training process at a point where the model generalized better to the unseen test data. The model trained without Early Stopping might have continued training beyond an optimal point, potentially starting to memorize the training data and thus performing slightly worse on the test set.

The use of `restore_best_weights=True` in the Early Stopping callback is crucial as it ensures that the model's weights are set back to the values that yielded the best performance on the validation set, rather than the weights at the very last epoch of training (which might have been suboptimal).